src/Orleans/Async/Cancellation/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<Guid, GrainCancellationTokenSource> _cancellationTokenSources;
```

Comment:
When the request with Orleans Token arrives, you pas that token to the the application.
When application asks for .NET Token out of Orleans Token, you can create the .NET Token Source, store it inside the Orleans Token and return to the application the .NET Token out of the .NET Token Source.

That way the Orleans Token source is merely application facade, you never really pass it around or store it. You only store all Orleans Tokens on the destination silo (and .NET Token Source inside it, if .NET Token was asked).

Makes sense?